File: src/pb.lua
--- 协议解析
module("pb", package.seeall)

--- 用于保存接收到的网络数据用于解析
local base = 2^8
local unique = 0
local data = nil
local pos = 1

--- 生成unique
local function genUnique()
    return 1
end
-- local className = ""

local function readbyte()
    local b1 = string.byte(data, pos)
    return b1
end

local function read2byte()
    local b1 = string.byte(data, pos)
    local b2 = string.byte(data, pos+1)
    pos = pos + 2
    return b1 * base + b2
end

local function readu4byte()
    local b1 = string.byte(data, pos)
    local b2 = string.byte(data, pos+1)
    local b3 = string.byte(data, pos+2)
    local b4 = string.byte(data, pos+3)
    if b1 and b2 and b3 and b4 then
        pos = pos + 4
        local value = b1*(2^24)+b2*(2^16)+b3*(2^8)+b4
        return value
    end
    assert("read4byte error")
end

local function writebyte(byte)
    local tp = type(byte)
    if tp == "number" then
        assert(byte < 256 and byte > -128, "byte lager then 256")
        return string.char((byte+256) % 256)
    elseif tp == "string" then
        return string.sub(byte, 1, 1)
    else
        -- error("writeByte type error")
        return string.char(0)
    end
end

local function write2byte(bytes)
    local tp = type(bytes)
    if tp == "number" then
        assert(bytes < (2 ^ 16), "bytes larger then 2^16")
        return string.char(math.floor(bytes / base)) .. string.char(math.floor(bytes % base))
    elseif tp == "string" then
        assert(#bytes>=2, "bytes len smaller then 2")
        return string.sub(bytes, 1, 2)
    else
        return string.char(0) .. string.char(0)
        -- error("write2byte type error")
    end
end

local function writeu4byte(bytes)
    local tp = type(bytes)
    if tp == "number" then
        assert(bytes < (2^32), "bytes larger then 2^32")
        local strTmp = ""
        for i=1, 4 do
            strTmp = strTmp .. string.char(bytes % base)
            bytes = math.floor(bytes / base)
        end
        return string.reverse(strTmp)
    elseif tp == "string" then
        assert(#bytes >= 4, "bytes len smaller then 4")
        return string.sub(bytes, 1, 4)
    else
        return string.rep(string.char(0), 4)
        -- error("write4byte type error")
    end
end

function pack(data)
    local fun = data.class
    local unique = genUnique()
    local methodID = pb_map.f2m[fun]
    local bin = pb_pack[fun](data)
    if #bin > 2048 then
        methodID = 2^31+methodID
        local compress = zlib.deflate()
        bin = compress(bin, "finish")
    end
    return write2byte(unique) .. writeu4byte(methodID) .. bin
end

function unpack(bin)
    data = bin
    pos = 1
     -- local unique = read2byte()
    local methodID = readu4byte()
    if methodID >= 2 ^ 31 then
        methodID = methodID - 2 ^ 31
        local uncom = zlib.inflate()
        data = uncom(string.sub(bin, 5))
        pos = 1
    end
    local fun = pb_map.m2f[methodID]
    print("begin unpack proto: ", methodID, fun)
    local tuple = {class = fun}
    pb_unpack[fun](tuple)
    return tuple
end

function rint32()
    local b1 = string.byte(data, pos)
    local b2 = string.byte(data, pos+1)
    local b3 = string.byte(data, pos+2)
    local b4 = string.byte(data, pos+3)
    if b1 and b2 and b3 and b4 then
        pos = pos + 4
        local value = b1*(2^24)+b2*(2^16)+b3*(2^8)+b4
        if b1 >= 2^7 then
            return -(2^32-value)
        else 
            return value
        end
    end
end

function rint64()
    local b1 = string.byte(data, pos)
    local sum = b1
    for i=2, 8 do
        sum = sum + string.byte(data, pos+i-1) * math.pow(2, (8-i)*8)
    end
    pos = pos + 8 
    if b1 >= 2^7 then
        return -(2^64 - sum)
    else
        return sum
    end
end

function rbool()
    local b1 = string.byte(data, pos)
    if b1 == 0 then
        return false
    else
        return true
    end
end

function rstring()
    local len = read2byte()
    if len == 0 then
        return ""
    end
    local str = string.sub(data, pos, pos+len-1)
    pos = pos + len 
    return str 
end

function rtuple(tupleName) 
    local tag = readbyte()
    if tag == 0 then
        return false
    end
    local tuple = {class=tupleName}
    pb_unpack[tupleName](tuple)
    return tuple
end

function rlist(tp)
    local len = read2byte()
    local list = {}
    local fun
    if tp == "int32" then
        fun = rint32
    elseif tp == "int64" then
        fun = rint64
    elseif tp == "bool" then
        fun = rbool
    elseif tp == "string" then
        fun = rstring
    else
        fun = rtuple
    end
    for i=1, len do 
        table.insert(list, fun(tp))
    end

    return list
end

function wint32(int)
    if not int then
        return string.rep(string.char(0), 4)
    end

    assert(int < 2^31 and int >= -2^31)
    if math.abs(int) ~= int then
        int = 2 ^ 32 + int 
    end
    local tmp = ""
    for i=1, 4 do 
        tmp = string.char(int % base) .. tmp
        int = math.floor(int / base)
    end
    return tmp
end

function wint64(int)
    if not int then
        return string.rep(string.char(0), 8)
    end 

    assert(int < 2^63 and int >= -2^63)
    if math.abs(int) ~= int then
        int = 2^64 + int 
    end
    local tmp = ""
    for i=1, 8 do 
        tmp = string.char(int % base) .. tmp
        int = math.floor(int/base)
    end
    return tmp
end

function wbool(bool)
    b = b and 1 or 0
    return string.char(b)
end

function wstring(str)
    if type(str) ~= "string" then
        str = ""
    end
    local len = #str 
    return (write2byte(len) .. str)
end

function wtuple(tuple)
    if not tuple then
        return writeByte(0)
    end
    local str = pb_pack[tuple.class](tuple)
    return writebyte(1) .. str
end

function wlist(list, tp)
    if not list then
        list = {}
    end
    local len = #list
    local str = write2byte(len)
    local fun 
    if tp == "int32" then
        fun = wint32
    elseif tp == "int64" then
        fun = wint64
    elseif tp == "bool" then
        fun = wbool
    elseif tp == "string" then
        fun = wstring
    else 
        fun = wtuple
    end
    for i=1, len do
        str = str .. fun(list[i], tp)
    end
    return str
end

File: src/pb.erl
%%%-------------------------------------------------------------------
%%% @author wanghaohao
%%% @copyright (C) 2015, <COMPANY>
%%% @doc
%%%
%%% @end
%%% Created : 08. 十月 2015 09:20
%%%-------------------------------------------------------------------
-module(pb).
-author("wanghaohao").

-define(ModuleID,(MethodID div 100)).

-include("common.hrl").
%% -include("all_pb.hrl").
%% -include("mm_define.hrl").

%% API
-export([unpack/1, pack/1]).
-export([rint32/1, rint64/1, rbool/1, rstring/1, rtuple/2, rlist/2, rdouble/1]).
-export([wint32/1, wint64/1, wbool/1, wstring/1, wtuple/2, wlist/2, wdouble/1]).

-compile(export_all).

-define( STRING(Str), (if is_list(Str) -> Str; is_binary(Str) -> binary_to_list(Str); true -> Str end)).

unpack(<<Unique:16, 1:1, MethodID:31, Bin/binary>>) ->
  Fun = pb_map:m_to_f(MethodID),
  {Tos, _} = pb_unpack:Fun(zlib:uncompress(Bin)),
  {Unique, ?ModuleID, MethodID, Tos};
unpack(<<Unique:16, MethodID:32, Bin/binary>>) ->
  Fun = pb_map:m_to_f(MethodID),
  {Tos, _} = pb_unpack:Fun(Bin),
  {Unique, ?ModuleID, MethodID, Tos}.

pack(R) ->
  Class = element(1, R),
  MethodID = pb_map:f_to_m(Class),
  Bin = pb_pack:Class(R),
  if
    byte_size(Bin) > 2048 ->
      Iscompress = 1,
      Bin1 = zlib:compress(Bin);
    true ->
      Iscompress = 0,
      Bin1 = Bin
  end,
  <<Iscompress:1, MethodID:31, Bin1/binary>>.

rint32(<<I:32/signed, B/binary>>) ->
  {I, B}.

rint64(<<I:64/signed, B/binary>>) ->
  {I, B}.

rdouble(B) ->
  rint64(B).

rbool(<<0:8, B/binary>>) ->
  {false, B};
rbool(<<_:8, B/binary>>) ->
  {true, B}.

rstring(<<Len:16, I:Len/binary-unit:8, B/binary>>) ->
  {?STRING(I), B}.

rtuple(<<0, B/binary>>, _Type) ->
  {undefined, B};
rtuple(<<1, B/binary>>, Type) ->
  pb_unpack:Type(B).

rlist(<<Len:16, B/binary>>, Type) ->
  {List,LeftBin} = case Type of
    int32 ->
      lists:foldl(fun(_, {Acc, Bin}) ->
        {I, Bin1} = rint32(Bin),
        {[I|Acc], Bin1}
      end, {[], B}, lists:seq(1, Len));
    int64 ->
      lists:foldl(fun(_, {Acc, Bin}) ->
        {I, Bin1} = rint64(Bin),
        {[I|Acc], Bin1}
      end, {[], B}, lists:seq(1, Len));
    bool ->
      lists:foldl(fun(_, {Acc, Bin}) ->
        {I, Bin1} = rbool(Bin),
        {[I|Acc], Bin1}
      end, {[], B}, lists:seq(1, Len));
    string ->
      lists:foldl(fun(_, {Acc, Bin}) ->
        {I, Bin1} = rstring(Bin),
        {[I|Acc], Bin1}
      end, {[], B}, lists:seq(1, Len));
    _ ->
      lists:foldl(fun(_, {Acc, Bin}) ->
        {I, Bin1} = rtuple(Bin, Type),
        {[I|Acc], Bin1}
      end, {[], B}, lists:seq(1, Len))
  end,
  {lists:reverse(List), LeftBin}.

wint32(V) when is_integer(V)->
  <<V:32/signed>>;
wint32(_) ->
  <<0:32/signed>>.

wint64(V) when is_integer(V) ->
  <<V:64/signed>>;
wint64(_) ->
  <<0:64/signed>>.

wdouble(V) ->
  wint64(V).

wbool(V) ->
  case V =:= false orelse V =:= 0 of
    true ->
      <<0>>;
    false ->
      <<1>>
  end.

wstring(V) ->
  Bin = if
    V =:= undefined -> <<>> ;
    is_list(V)      -> list_to_binary(V);
    is_integer(V)   -> list_to_binary(integer_to_list(V));
    is_atom(V)      -> atom_to_binary(V,latin1);
    is_tuple(V)     -> list_to_binary(tuple_to_list(V));
    is_bitstring(V) -> list_to_binary(bitstring_to_list(V));
    true               -> V
  end,
  <<(size(Bin)):16,Bin/binary>>.

wtuple(undefined, _Type) ->
  <<0>>;
wtuple(V, Type) ->
  <<1,(pb_pack:Type(V))/binary>>.




wlist(undefined, Type) ->
  wlist([], Type);
wlist(List, Type) when is_list(List) ->
  Len = length(List),
  Bin = lists:foldl(fun(V, Acc) ->
    <<Acc/binary, (witem(V, Type))/binary>>
  end, <<>>, List),
  <<Len:16, Bin/binary>>.

witem(V, Type) ->
  case Type of
    int32 ->
      wint32(V);
    int64 ->
      wint64(V);
    double ->
      wint64(V);
    bool ->
      wbool(V);
    string ->
      wstring(V);
    _ ->
      wtuple(V, Type)
  end.




















